packages/salesforcedx-vscode-core/src/commands/util/sfdxCommandlet.ts

```diff
@@ -95,28 +111,74 @@ export abstract class SfdxCommandletExecutor<T>
     });
 
     execution.processExitSubject.subscribe(exitCode => {
-      const telemetryData = this.getTelemetryData(
-        exitCode === 0,
-        response,
-        output
-      );
-      let properties;
-      let measurements;
-      if (telemetryData) {
-        properties = telemetryData.properties;
-        measurements = telemetryData.measurements;
-      }
-      this.logMetric(
-        execution.command.logName,
-        startTime,
-        properties,
-        measurements
-      );
-      this.onDidFinishExecutionEventEmitter.fire(startTime);
+      this.exitProcessHandler(exitCode, execution, response, startTime, output);
     });
+
     this.attachExecution(execution, cancellationTokenSource, cancellationToken);
   }
 
+  protected exitProcessHandler(
+    exitCode: number | undefined,
+    execution: CliCommandExecution,
+    response: ContinueResponse<T>,
+    startTime: [number, number],
+    output: string
+  ): void {
+    if (
+      exitCode === 0 &&
+      execution.command.logName === FORCE_SOURCE_PULL_LOG_NAME
+    ) {
+      const pullResult = this.parseOutput(output);
+      this.updateCache(pullResult);
+
+      const pullParser = new ForcePullResultParser(output);
+      const errors = pullParser.getErrors();
+      if (errors) {
+        channelService.showChannelOutput();
+      }
+
+      this.outputResultPull(pullParser);
+    }
+
+    const telemetryData = this.getTelemetryData(
+      exitCode === 0,
+      response,
+      output
+    );
+    let properties;
+    let measurements;
+    if (telemetryData) {
+      properties = telemetryData.properties;
+      measurements = telemetryData.measurements;
+    }
+    this.logMetric(
+      execution.command.logName,
+      startTime,
+      properties,
+      measurements
+    );
+    this.onDidFinishExecutionEventEmitter.fire(startTime);
+  }
+
+  /**
+   * @param output usually stdOut JSON string from a cli command.
+   * @returns parsed JSON object. Shows a notification and the raw
+   * response before throwing when a parsing error is encountered.
+   */
+  protected parseOutput(output: string) {
+    let parsed: any;
+    try {
+      parsed = JSON.parse(output);
+    } catch (error) {
+      notificationService.showWarningMessage(
+        nls.localize('lib_retrieve_result_parse_error'),
+        output
+      );
+      throw error;
+    }
+    return parsed;
+  }
```

Comment:
This method now makes it look like this if there is a parsing error - let me know what you think
![2023-05-19_17-13-34 (1)](https://github.com/forcedotcom/salesforcedx-vscode/assets/46458081/7fc999c0-82a8-48b7-ba87-d34d2646d2dd)